File: Android/HatPepper/app/src/main/java/jp/nuits/hatpepper/infrastructures/GourmetSearchApiImpl.java
package jp.nuits.hatpepper.infrastructures;

import java.util.function.Consumer;

import jp.nuits.hatpepper.entities.GourmetSearchApi;
import jp.nuits.hatpepper.entities.GourmetSearchResults;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class GourmetSearchApiImpl implements GourmetSearchApi {

    private final Retrofit retrofit;

    public GourmetSearchApiImpl(Retrofit retrofit) {
        this.retrofit = retrofit;
    }

    @Override
    public void search(Consumer<GourmetSearchResults> consumer) {
        GourmetRetrofitService retrofitService = retrofit.create(GourmetRetrofitService.class);
        Call<GourmetSearchResults> call = retrofitService.Search(Secrets.API_KEY);
        call.enqueue(new Callback<GourmetSearchResults>() {
            @Override
            public void onResponse(Call<GourmetSearchResults> call, retrofit2.Response<GourmetSearchResults> response) {
                GourmetSearchResults gourmetSearchResults = response.body();
                consumer.accept(gourmetSearchResults);
            }

            @Override
            public void onFailure(Call<GourmetSearchResults> call, Throwable t) {
                // 本来は例外処理を行うべきだが、本題から外れるため省略
            }
        });
    }
}

File: Android/01.Before/infrastructure_api/src/main/java/jp/nuits/hatpepper/infrastructure/api/GourmetSearchApiImpl.kt
package jp.nuits.hatpepper.infrastructure.api

import android.location.Location
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class GourmetSearchApiImpl : GourmetSearchApi {

    private val retrofit by lazy {
        Retrofit
            .Builder()
            .baseUrl("https://webservice.recruit.co.jp/")
            .addConverterFactory(GsonConverterFactory.create())
            .build() }

    override suspend fun find(location: Location, lunchTimeOnly : Boolean) : GourmetSearchResults {
        val retrofitService = retrofit.create(RetrofitGourmetSearchApi::class.java)
        val response =
            retrofitService.search(
                Secrets.API_KEY,
                location.longitude,
                location.latitude,
                if(lunchTimeOnly) 1 else 0)
        return response.body()!!
    }
}